### PR TITLE
Support normalized project names

### PR DESCRIPTION
**Pull request by ofek**
- https://www.python.org/dev/peps/pep-0621/#name
- https://www.python.org/dev/peps/pep-0503/#normalized-names
- https://www.python.org/dev/peps/pep-0508/#names

### CODE REVIEW

File: src/editables/__init__.py [hunk 1]
```diff
@@ -15,6 +15,7 @@ class EditableException(Exception):
 class EditableProject:
     def __init__(self, project_name, project_dir):
         self.project_name = project_name
+        self.package_name = project_name.replace('-', '_')
```
Comment (ofek) [inline src/editables/__init__.py:18]:
```suggestion
        self.package_name = project_name.replace("-", "_")
```

**Comment by pfmoore** [comment]
Hmm. I never really considered normalising here, because I wasn't thinking of the `project_name` argument as having to be a "project name" in the packaging sense, just an arbitrary name that would be used for the `.pth` and bootstrap files. Maybe the argument name should be changed to be more generic?

One thing I definitely want is to use the *same* name for the `.pth` file and the bootstrap file, so there's no need for a separate `package_name` attribute.

I'd rather not assume that the argument is a normalised project name in the packaging sense. Either we accept any arbitrary name, and normalise it ourselves (using a rule of "PEP 503 with dashes replaced by underscores in the final result", i.e., `re.sub(r"[-_.]+", "_", project_name).lower()`, I guess), or we require that the name passed in must be a valid name *for an importable Python file* (not for a Python project) and I'd be fine with adding some validation if we did that.

Other notes:

1. The PR needs tests adding.
2. Coverage should still be at 100% with the PR.
3. The PR should probably also add some clarification to the documentation, explaining what a valid argument is here. If you misunderstood it as being the name of the project in a packaging sense, then so could others.

One thing to consider is there's no requirement that a single project would only include the results of one call to `EditableProject`. It's the usual and expected use case, but you might want more than one - so tying the argument of `EditableProject("whatever")` to the name of the project a wheel is being built for is unnecessarily restrictive. I can't see why people would need to do that, but I also see no reason to stop them if they *do* find a good reason for doing so.

**Comment by pfmoore** [comment]
@ofek do you have any interest in taking this PR further?

**Comment by ofek** [comment]
Sorry, updated!

**Comment by ofek** [comment]
(Let me know if you'd just prefer a doc update instead)

**Comment by pfmoore** [comment]
I think we need to have a discussion about what's needed here, and actually design things correctly. I see you noticed that your initial approach caused the tests to fail because the name was no longer a valid Python variable name. I'm not keen on your fix, either, I'd rather we stuck as close to existing normalisation rules as we can.

I think we need to establish some usage principles here:

1. The user is creating an `EditableProject` object for a particular project, which has a "project name" as part of the package metadata. We'll call the metadata name the "project name" and say that's what we expect to have passed to us (but see below). 2. The project name, does not need to be normalised when passed to us, but we will normalise it before using it. The PEP 503 normalised form is *not* necessarily a valid Python import name, though, so we need to address that.
3. The user won't create multiple `EditableProject` objects for the same project (although we must state clearly what will happen if they violate this assumption).
4. The project could include importable modules with arbitrary names. Project `foo` could expose `foo.py`, `_foo.py`, or anything else.
5. The support files we generate should have discoverable names, so that someone looking in site-packages can make a reasonable guess as to what the files are for. So no randomly generated filenames...

With those constraints in mind, I'd suggest that we:

1. Take the supplied project name, normalise it, and then convert `-` characters to underscores (I'm pretty sure that gives a valid Python variable name in all cases, but we should check). We document that the project name *must not* clash with existing names (for example, Python reserved words, stdlib modules, ...). I don't want to invent yet another variation on normalisation rules here, so I'd rather stick to "PEP 503 with dashes replaced by underscores" as the minimal variation that's workable.
2. Generate files `_{name}_editable_support.py` and `{name}_editable_support.pth`. Document that we reserve names of the form `{name}*_editable_support*` for internal use, and it's not supported for projects to use those names themselves.
3. Document that callers must *not* create multiple `EditableProject` objects for the same project name - if they have an unusual requirement that means they need to do something like that, we recommend adding a numeric suffix, so project `foo` passes project names of `foo_1`, `foo_2`, ...

As for tests, I did mean *add* some tests, not just use a different project name that needs normalising 😉 I'd rather that you took a test-driven approach here, and added some tests that fail with the current code, demonstrating the problem, and are fixed by the new code. Leave all of the existing tests unchanged. It might even be good to have a parametrised test that checks that everything works for a range of tricky project names.

And finally, the docs do need updating. But proper docs need writing in the first place - there's only so much detail you can cover in a README 🙁 And that's on me. I'll make sure this gets included when I finally get around to doing the docs properly.

**Comment by pfmoore** [comment]
@ofek sorry for the delay on this. I finally got round to writing some docs, and picked this up at the same time. In the end, I took your ideas and did my own implementation, which is at #18. Would you be able to take a look and tell me what you think?

**Comment by ofek** [comment]
That looks great, thanks!!!

**Pull request closed without merging**